docs/taskman-design.md

```diff
@@ -0,0 +1,379 @@
+
+```
+tlsConfig := &tls.Config{
+	Certificates: []tls.Certificate{serverCert},
+	ClientAuth:   tls.RequireAndVerifyClientCert,
+	ClientCAs:    caCertPool,
+	MinVersion:   tls.VersionTLS13,
+	CurvePreferences: []tls.CurveID{
+        tls.X25519,
+        tls.CurveP256,
+        tls.CurveP384,
+    },
+}
+```
+Since only TLS version 1.3 is being supported we do not need to include a list of CipherSuites because they are fixed in Go see [crypto/tls](https://github.com/golang/go/blob/master/src/crypto/tls/common.go#L688-L697).
+
+The client and server certificates will be signed by a CA that is generated locally. This CA certificate will be self signed and not by a trusted third party. These certificates will be committed to the exercise repository as a proof of concept. These certificates will be generated using `openssl` with the following details:
+
+```
+Public Key Algorithm: rsaEncryption
+Public Key Size: 4096 bit
+Signature Algorithm: sha256WithRSAEncryption
+```
+
+In a production environment, we would want these certificates to be generated, short-lived, and renewed through automation securely.
+
+#### Authorization
+Each certificate will be generated with a Common Name (CN) that will identify the user. The certificate's common name will be extracted by the gRPC server and associated with each created task. When a user makes a request against a task their common name will be extracted from the provided certificate on the request and must match the name associated with the task on the server in order for requests to be accepted.
+
+There will be an admin certificate that the gRPC server allows access to all tasks and operations on the server.
+
+### Privacy
+The gRPC server will only store data in memory for the life of the server and will be lost on server shutdown or restart. Any other privacy considerations are beyond the scope of this exercise.
+
+### Proto Specification
+This exercise will have a new protobuf representing the TaskManager service that exposes functionality to start, stop, get, and stream tasks.
+
+```package task_manager;
+
+// TaskManager is a service that exposes APIs to manage linux processes (tasks)
+service TaskManager {
+    // StartTask takes a task and arguments and starts a new task
+    rpc StartTask (StartTaskRequest) returns (StartTaskResponse);
+    // StopTask stops a running task by task ID
+    rpc StopTask (StopTaskRequest) returns (StopTaskResponse);
+    // GetTaskStatus gets the status of a task by task ID
+    rpc GetTaskStatus (TaskStatusRequest) returns (TaskStatusResponse);
+    // StreamTaskOutput streams the output of a task by task ID
+    rpc StreamTaskOutput (StreamTaskOutputRequest) returns (stream StreamTaskOutputResponse);
+}
+
+// JobStatus tracks status of job
+enum JobStatus {
+    // job is in an unknown error state
+    JOB_STATUS_UNKNOWN = 0;
+    // job is currently running
+    JOB_STATUS_STARTED = 1;
+    // job was manually stopped by client
+    JOB_STATUS_STOPPED = 2;
```

Comment:
To add to @rosstimothy's comment a bit, will this state apply to jobs stopped only by `taskman` via user request, or also to jobs killed via signal? (`kill -9`, OOM kill, etc).

Reply:
We could get more granular and allow more information back to the user if we change this status to be `JOB_STATUS_SIGNALED` and include in the status object
```
string signal = 8;         // e.g., "SIGTERM, SIGKILL"
string termination_source = 9; // "user", "system", "oom", etc.
```

Would this approach be preffered?